src/main/scala/org/ergoplatform/nodeView/history/extra/ExtraIndexer.scala

```diff
@@ -0,0 +1,473 @@
+      globalTxIndex -= 1
+    }
+    globalTxIndex += 1
+    historyStorage.removeExtra(txs.toArray)
+
+    // remove all box indexes, tokens and address balances
+    val boxTarget: Long = history.typedExtraIndexById[IndexedErgoBox](bytesToId(lastTxToKeep.outputs.last.id)).get.globalIndex
+    val toRemove: ArrayBuffer[ModifierId] = ArrayBuffer.empty[ModifierId]
+    globalBoxIndex -= 1
+    while(globalBoxIndex > boxTarget) {
+      val iEb: IndexedErgoBox = NumericBoxIndex.getBoxByNumber(history, globalBoxIndex).get
+      val address: IndexedErgoAddress = history.typedExtraIndexById[IndexedErgoAddress](bytesToId(hashErgoTree(iEb.box.ergoTree))).get
+      address.spendBox(iEb.box)
+      if(tokenRegistersSet(iEb.box))
+        history.typedExtraIndexById[IndexedToken](IndexedTokenSerializer.fromBox(iEb.box).id) match {
+          case Some(token) => toRemove += token.id // token created, delete
+          case None => // no token created
+        }
+      address.rollback(txTarget, boxTarget)(_history)
+      toRemove += iEb.id // box by id
+      toRemove += bytesToId(NumericBoxIndex.indexToBytes(globalBoxIndex)) // box id by number
+      globalBoxIndex -= 1
+    }
+    globalBoxIndex += 1
+    historyStorage.removeExtra(toRemove.toArray)
+
+    saveProgress(false)
+
+    log.info(s"Successfully rolled back indexes to $height")
+
+    rollback = false
+
+  }
+
+}
+
+
+/**
+  * Actor that constructs an index of database elements.
+  * @param cacheSettings - cacheSettings to use for saveLimit size
+  */
+class ExtraIndexer(cacheSettings: CacheSettings)
+  extends Actor with ExtraIndexerBase {
+
+  override val saveLimit: Int = cacheSettings.history.extraCacheSize * 10
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[FullBlockApplied])
+    context.system.eventStream.subscribe(self, classOf[Rollback])
+  }
+
+  override def postStop(): Unit =
+    log.info(s"Stopped extra indexer at height ${if(lastWroteToDB > 0) lastWroteToDB else indexedHeight}")
+
+  override def receive: Receive = {
+
+    case FullBlockApplied(header: Header) if caughtUp =>
+      index(history.typedModifierById[BlockTransactions](header.transactionsId).get, header.height) // after the indexer caught up with the chain, stay up to date
+
+    case Rollback(branchPoint: ModifierId) =>
+      val branchHeight: Int = history.heightOf(branchPoint).get
+      rollback = branchHeight < indexedHeight
+      if(rollback) {
+        removeAfter(branchHeight)
+        run() // restart indexer
+      }
+
+    case Start(history: ErgoHistory) =>
+      _history = history
+      run()
+
+  }
+}
+
+object ExtraIndexerRef {
+
+  object ReceivableMessages {
+    /**
+      * Initialize ExtraIndexer and start indexing.
+      * @param history - handle to database
+      */
+    case class Start(history: ErgoHistory)
+  }
+
+  private var _ae: ErgoAddressEncoder = _
```

Comment:
var in object does not look good, please rework if possible

Reply:
done